velox/type/Filter.cpp

```diff
@@ -918,6 +935,115 @@ bool BigintValuesUsingHashTable::testInt64Range(
   return max >= *it;
 }
 
+// Follows BigintValuesUsingHashTable for hash table sizing and padding.
+HugeintValuesUsingHashTable::HugeintValuesUsingHashTable(
+    const int128_t min,
+    const int128_t max,
+    const std::vector<int128_t>& values,
+    const bool nullAllowed)
+    : Filter(true, nullAllowed, FilterKind::kHugeintValuesUsingHashTable),
+      min_(min),
+      max_(max) {
+  constexpr int32_t kPaddingElements = 4;
+  VELOX_CHECK(min <= max, "min must not be greater than max");
+  VELOX_CHECK(!values.empty(), "values must not be empty");
+
+  auto size = 1u << (uint32_t)std::log2(values.size() * 5);
+  hashTable_.resize(size + kPaddingElements);
+  sizeMask_ = size - 1;
+  std::fill(hashTable_.begin(), hashTable_.end(), kEmptyMarker);
+  for (auto value : values) {
+    if (value == kEmptyMarker) {
+      containsEmptyMarker_ = true;
+    } else {
+      uint32_t position = ((value * M) & sizeMask_);
+      for (auto i = position; i < position + size; i++) {
+        uint32_t index = i & sizeMask_;
+        if (hashTable_[index] == kEmptyMarker) {
+          hashTable_[index] = value;
+          break;
+        }
+      }
+    }
+  }
+
+  for (auto i = 0; i < kPaddingElements; ++i) {
+    hashTable_[sizeMask_ + 1 + i] = hashTable_[sizeMask_];
+  }
+  std::sort(values_.begin(), values_.end());
+}
+
+bool HugeintValuesUsingHashTable::testInt128(const int128_t value) const {
+  if (containsEmptyMarker_ && value == kEmptyMarker) {
+    return true;
+  }
+  if (value < min_ || value > max_) {
+    return false;
+  }
+  uint32_t pos = (value * M) & sizeMask_;
+  for (auto i = pos; i <= pos + sizeMask_; i++) {
+    uint32_t idx = i & sizeMask_;
+    int128_t l = hashTable_[idx];
+    if (l == kEmptyMarker) {
+      return false;
+    }
+    if (l == value) {
+      return true;
+    }
+  }
+  return false;
+}
+
+folly::dynamic HugeintValuesUsingHashTable::serialize() const {
+  auto obj = Filter::serializeBase("HugeintValuesUsingHashTable");
+  obj["min_lower"] = HugeInt::lower(min_);
```

Comment:
Can't we serialize int128_t values?

Reply:
+1

Reply:
Thanks for the suggestion, I checked ```asImpl() in folly/dynamic-inl.h``` before and could not find a function to serialize int128 values. Will check if there are other folly functions to serialize int128 values and update here.